Document debugging in RustRover with Dynamic Linking (#376)
TheBevyFlock/bevy_new_2d: docs/tooling.md

To make it easier to enable or disable these live templates for different projects, you can put them in a template group called `Bevy`.

## Debugging with RustRover  

The Template includes a Cargo Run Configuration for debugging purposes, however it has `--no-default-features` so it will work without any user configuration. However, this disables Dynamic Linking when using
that option, along with several other tools and features that are provided for development/debugging purposes. If you want to be able to use these features while debugging you can do the following to get RustRover
(or CLion or IntelliJ IDEA Ultimate) to debug a bevy game with dynamic linking enabled:

1. Open the terminal and type in the following: `rustc --print target-libdir` and copy the output.

> 2. GO to edit the "Run Native Debug" run configuration that is available within the IDE.

Reviewer: Note to *not* click on the shell scripts.

3. Add in the following Environment Variable:

>   a. On Linux or Mac set the name of the Environment Variable to `LD_LIBRARY_PATH` with the following value: `./target/debug/deps:[LIBDIR_PATH]` where `LIBDIR_PATH` is the output of the command in step 1.

Reviewer: Add note about being careful about whitespace

To make it easier to enable or disable these live templates for different projects, you can put them in a template group called `Bevy`.

## Debugging with RustRover  

This template comes with a Cargo Run Configuration that disables dynamic linking (and dev tools) so that the debugger will work out of the box. If you'd like to enable those features in the debugger, it'll require some setup:

1. Run `rustc --print target-libdir` and copy the output. You can specify a channel here with e.g. `rustc +nightly --print target-libdir`
2. Go to edit the "Run Native Debug" run configuration that is available within the IDE, this should be a Cargo Run Configuration.
  - If the configuration displays a little terminal icon, then it is not the right Run Configuration.

>   - If you do not have this -- or any Cargo Run Configuration -- you can make one by clicking on the current Configuration, clicking Edit Configurations, and then clicking the Plus Icon and then Cargo.

Author: I mostly put this here because Jan tried doing this before that got merged, and mostly just in-case they somehow ended up without it. Although I'm fine removing it.

>   - If you do not have this -- or any Cargo Run Configuration -- you can make one by clicking on the current Configuration, clicking Edit Configurations, and then clicking the Plus Icon and then Cargo.

Reviewer 2: I see, should be good to remove it then since 99% of users will be on a version of the template that includes the `.idea/` if they're reading the updated docs.

3. Add in the following Environment Variable:
  a. Linux or Mac: `LD_LIBRARY_PATH` = `./target/debug/deps:<LIBDIR_PATH>` where `<LIBDIR_PATH>` is the output from step 1.
  b. Windows: `PATH` = `.\target\debug\deps:<LIBDIR_PATH>`, where `<LIBDIR_PATH>` is the output from step 1.
3. Remove the `--no-default-features` from the command line argument in the Run Configuration.
4. Click Apply and then Debug, and if everything is correct it should launch the game.

If you want to use multiple different channels for the same project, you will need to add in a `LIBDIR_PATH` for every channel you intend on using.

If you're still having issues, please ensure that the channels in the path and the Run Configuration match, and that there are no extra spaces (especially at the beginning or end).

> [!NOTE]
> <details>

> > <summary>About the steps provided in this section</summary>
> >
> > This is not the only way to debug a game in RustRover with Dynamic Linking enabled. If you are using the Shell Script Run Configurations, you can attach the debugger to the game afterwards.
> > To do this, you will need to run the Shell Script Run Configuration you want to use, then use Run > Attach to Process and attach the process that is the name of your game, not the one named `bevy`.
> >
> > This does not work for the web scripts.

Reviewer 2: ```suggestion
> <summary>Attaching the debugger to a running game</summary>
>
> If you started your game with a Shell Script Run Configuration, you can attach the debugger to it while it's running by using `Run > Attach to Process` and selecting the process with the same name as your game (not the one named `bevy`).
>
> This does not work for web builds.
```

To make it easier to enable or disable these live templates for different projects, you can put them in a template group called `Bevy`.

## Debugging with RustRover  

> The Template includes a Cargo Run Configuration for debugging purposes, however it has `--no-default-features` so it will work without any user configuration. However, this disables Dynamic Linking when using
> that option, along with several other tools and features that are provided for development/debugging purposes. If you want to be able to use these features while debugging you can do the following to get RustRover
> (or CLion or IntelliJ IDEA Ultimate) to debug a bevy game with dynamic linking enabled:

Reviewer 2: ```suggestion
This template comes with a Run Configuration that disables dynamic linking (and dev tools) so that the debugger will work out of the box. If you'd like to enable those features in the debugger, it'll require some setup:
```

> 1. Open the terminal and type in the following: `rustc --print target-libdir` and copy the output.

Reviewer 2: ```suggestion
1. Run `rustc --print target-libdir` and copy the output. You can specify a channel here with e.g. `rustc +nightly --print target-libdir`.
```

Author: Thank-you I actually didn't know specifying the channel was entirely possible

2. Go to edit the "Run Native Debug" run configuration that is available within the IDE, this should be a Cargo Run Configuration.

>   - If you do not have this -- or any Cargo Run Configuration -- you can make one by clicking on the current Configuration, clicking Edit Configurations, and then clicking the Plus Icon and then Cargo.

Reviewer: ```suggestion
  - If the configuration has a little terminal icon, it is the wrong kind of configuration
  - If you do not have this -- or any Cargo Run Configuration -- you can make one by clicking on the current Configuration, clicking Edit Configurations, and then clicking the Plus Icon and then Cargo.
```

3. Add in the following Environment Variable:

>   a. On Linux or Mac set the name of the Environment Variable to `LD_LIBRARY_PATH` with the following value: `./target/debug/deps:[LIBDIR_PATH]` where `LIBDIR_PATH` is the output of the command in step 1.

Reviewer 2: ```suggestion
  a. Linux or Mac: `LD_LIBRARY_PATH` = `./target/debug/deps:<LIBDIR_PATH>` where `<LIBDIR_PATH>` is the output from step 1.
```

>   b. On Windows set the name of the Environment Variable to `PATH` with the following value: `.\target\debug\deps:[LIBDIR_PATH]` where `LIBDIR_PATH` is the output of the command in step 1.

Reviewer 2: ```suggestion
  b. Windows: `PATH` = `.\target\debug\deps:<LIBDIR_PATH>`, where `<LIBDIR_PATH>` is the output from step 1.
```

> 3. Remove the `--no-default-features` part of the command line argument in the Run Configuration.

Reviewer 2: ```suggestion
3. Remove `--no-default-features` from the command in the Run Configuration.
```

> 4. Click on Apply, and then click Debug, if everything is correct it should launch the game.

Reviewer 2: ```suggestion
4. Click Apply and then Debug, and if everything is correct it should launch the game.
```

> After doing these steps you can now debug within RustRover, while keeping Dynamic Linking enabled, unfortunately because the path to your rust stdlib is dependent upon your own system. RustRover (and CLion and IntelliJ IDEA Ultimate)
> technically has Path Variables which *should* allow you to set these things, but it doesn't work with Cargo Run Configurations in Environment Variables apparently.
> 

Reviewer 2: ```suggestion
```
I'd prefer to record this information in an issue (new or existing).

> If you want to use a different channel (IE: Nightly instead of Stable or vice-versa) you will need to add in the path to that libdir, but that's as simple as replacing `stable` with `nightly` in the path you get from step 1,
> you can also add both if you want to be able to switch between channel, for whatever reason, you can add the stdlib paths for every channel you will be using.

Reviewer 2: ```suggestion
```
Moved this information to step 1 instead.

> Please note that if you do these steps *and* it still doesn't work, please ensure that you're on the right channel in the Run Configuration and in your libdir path, as they should match. (IE: Nightly, Stable, Beta), and also make
> sure that there are no erronious spaces -- especially at the end -- as any extra spaces (at the beginning or end) of the values *WILL* cause it to fail.

Reviewer 2: ```suggestion
If you're still having issues, please ensure that the channels in the path and the Run Configuration match, and that there are no extra spaces (especially at the beginning or end).
```

To make it easier to enable or disable these live templates for different projects, you can put them in a template group called `Bevy`.

## Debugging with RustRover  

This template comes with a Cargo Run Configuration that disables dynamic linking (and dev tools) so that the debugger will work out of the box. If you'd like to enable those features in the debugger, it'll require some setup:

1. Run `rustc --print target-libdir` and copy the output. You can specify a channel here with e.g. `rustc +nightly --print target-libdir`
2. Edit the Cargo Run Configuration named "Run Native Debug" (it should be the one without a terminal icon).
3. Add the following Environment Variable:
  a. Linux or Mac: `LD_LIBRARY_PATH` = `./target/debug/deps:<LIBDIR_PATH>` where `<LIBDIR_PATH>` is the output from step 1.
  b. Windows: `PATH` = `.\target\debug\deps:<LIBDIR_PATH>`, where `<LIBDIR_PATH>` is the output from step 1.

> 3. Remove the `--no-default-features` from the command line argument in the Run Configuration.

Reviewer 2: ```suggestion
3. Remove `--no-default-features` from the command in the Run Configuration.
```
